File: backend/app/services/pipeline/stages/analyze.py
"""Stage 2: Analyze code for security issues."""
from app.services.pipeline.base import PipelineContext
from app.services.analyzers.factory import create_analyzers
from app.services.rag_service import RagService


class AnalyzeStage:
    """Run security analyzers on diff files."""

    def execute(self, context: PipelineContext) -> PipelineContext:
        """Analyze files and collect findings."""
        rag = RagService()

        for file_item in context.files:
            path = file_item.get("filename")
            patch = file_item.get("patch")

            if not path or not patch:
                continue

            # Get RAG context
            rag_query = f"{path}\n{patch[:1500]}"
            rag_text = rag.retrieve(query_text=rag_query)

            # Create analyzers with RAG context
            analyzers = create_analyzers(rag_context=rag_text)

            # Run all analyzers
            for analyzer in analyzers:
                findings = analyzer.analyze(path, patch)
                context.findings.extend(findings)

        return context

    def get_name(self) -> str:
        return "AnalyzeStage"

File: backend/src/services/pipeline/stages/analyze.ts
import { PipelineContext, PipelineStage } from '../base.js';
import { createAnalyzers } from '../../analyzers/factory.js';
import { RagService } from '../../rag-service.js';
import { buildRagQueries } from '../../rag-query.js';

/**
 * Stage 2: Run security analyzers on diff files.
 */
export class AnalyzeStage implements PipelineStage {
  async execute(context: PipelineContext): Promise<PipelineContext> {
    const rag = new RagService();

    for (const fileItem of context.files) {
      const path = fileItem.filename as string | undefined;
      const patch = fileItem.patch as string | undefined;

      if (!path || !patch) {
        continue;
      }

      // Get RAG context. retrieve() never throws: when no relevant policy is
      // available it returns an empty context plus the reason, and the review
      // continues on general secure-coding knowledge.
      const ragContext = await rag.retrieve(buildRagQueries(path, patch));
      context.ragStatuses.push(ragContext.status);
      if (ragContext.status !== 'ok') {
        console.log(`[RAG] No policy context for ${path} (${ragContext.status})`);
      }

      // Create analyzers with RAG context
      const analyzers = createAnalyzers(ragContext);

      // Run all analyzers
      for (const analyzer of analyzers) {
        const findings = await analyzer.analyze(path, patch);
        context.findings.push(...findings);
      }
    }

    return context;
  }

  getName(): string {
    return 'AnalyzeStage';
  }
}
